### ISSUE TITLE
esbuild fails with transitive dependencies

### ISSUE THREAD

**Issue by paullewis** [aspect-build/rules_esbuild#37]
Hey,

I couldn't decide if this is an issue with the esbuild rules, npm_link_all_packages, or how I'm working, so please let me know if I've got this wrong.

If I have a `BUILD.bazel` file in the root like this:

```bazel
load("@aspect_rules_esbuild//esbuild:defs.bzl", "esbuild")
load("@npm//:defs.bzl", "npm_link_all_packages")

npm_link_all_packages(name = "node_modules")

esbuild(
    name = "index",
    entry_point = "index.ts",
    deps = [
        ":node_modules/http-errors",
    ],
)
```

Here `index.ts` is just importing `http-errors` as you'd typically expect, `http-errors` is listed in the `package.json`, and everything has been installed through `pnpm`.

When I try and build, however, esbuild is unable to find the dependencies of `http-errors`: `depd`, `setprototypeof`, `statuses`, `inherits`, and `toidentifier`, and fails:

```
✘ [ERROR] Could not resolve "depd"

    node_modules/http-errors/index.js:15:24:
      15 │ var deprecate = require('depd')('http-errors')
         ╵                         ~~~~~~
```

If I were to add the dependencies to the `deps` attribute, `:node_modules/depd`, I get another error because `npm_link_all_packages` only linked `http-errors` and not _its_ dependencies.

Not sure what the right thing is here. One option would be to add all transitive dependencies to the `package.json` file, but that creates a lot of overhead during dependency upgrades if versions get out of line. I guess another option is to mark them all as `external`, but there again that would create a lot of overhead to track deps of deps.

Is there a recommended way to do this?

**Comment by paullewis** [comment]
I just realised I can hoist the dependencies with `public_hoist_packages`, which I can use as a workaround. Overall, though, that seems like something to avoid (from a pnpm perspective at least), so I guess I'm still back at the original question as to whether there's a better way.

**Comment by jbedard** [comment]
It does declare the dependency though...

**Comment by paullewis** [comment]
Absolutely agree, that makes sense. The challenge is that `http-errors` does declare all the dependencies in its `package.json` file. It makes sense that it can't build because the sandbox only has access to `http-errors`, but I can't present _its_ dependencies in the sandbox, either directly or indirectly, unless I hoist. Not sure where the bug resides though!

**Comment by alexeagle** [comment]
Okay, since it does declare the dependencies properly, we can run

```
% bazel build :node_modules/http-errors

Target //:node_modules/http-errors up-to-date:
  bazel-bin/node_modules/.aspect_rules_js/http-errors@2.0.0/node_modules/http-errors
  bazel-bin/node_modules/.aspect_rules_js/http-errors@2.0.0/node_modules/depd
  bazel-bin/node_modules/.aspect_rules_js/http-errors@2.0.0/node_modules/inherits
  bazel-bin/node_modules/.aspect_rules_js/http-errors@2.0.0/node_modules/setprototypeof
  bazel-bin/node_modules/.aspect_rules_js/http-errors@2.0.0/node_modules/statuses
  bazel-bin/node_modules/.aspect_rules_js/http-errors@2.0.0/node_modules/toidentifier
  bazel-bin/node_modules/.aspect_rules_js/depd@2.0.0/node_modules/depd
  bazel-bin/node_modules/.aspect_rules_js/inherits@2.0.4/node_modules/inherits
  bazel-bin/node_modules/.aspect_rules_js/setprototypeof@1.2.0/node_modules/setprototypeof
  bazel-bin/node_modules/.aspect_rules_js/statuses@2.0.1/node_modules/statuses
  bazel-bin/node_modules/.aspect_rules_js/toidentifier@1.0.1/node_modules/toidentifier
  bazel-bin/node_modules/http-errors

% ls bazel-bin/node_modules/.aspect_rules_js/http-errors@2.0.0/node_modules/depd
History.md      Readme.md       lib
LICENSE         index.js        package.json
```

so what's supposed to happen here is that the require from `node_modules/http-errors/index.js` resolves to that folder.

**Comment by paullewis** [comment]
Right, yep, that works for me, too, I can build `:node_modules/http-errors` directly. Also if I run `esbuild` directly, too, i.e., `esbuild --bundle index.ts` outside of Bazel that works. However, if I try and compile via the `rules_esbuild` `esbuild` rule it fails because the dependencies of `http-errors` aren't available. I guess I don't fully understand exactly how dependencies are exposed to the sandbox (specifically the linking of packages under – say – `//:node_modules`).

I've pushed a repo with the setup discussed to https://github.com/paullewis/repro-rules-esbuild-issue-37, which I hope will be useful.

**Comment by mattem** [comment]
It breaks due to the preserve symlinks option being set as a hangover from rules_nodejs. However unsettling it currently causes other issues. 

https://github.com/aspect-build/rules_esbuild/pull/32
Needs pushing over the finish line.

**Comment by apesternikov** [comment]
I just want to confirm that #32 fixes this issue for me

**Comment by arobinson** [comment]
Is there a work-around for this? I'm currently blocked by the same problem. `esbuild` appears to work fine with `pnpm` outside of Bazel. Setting `preserveSymlinks` to true causes `ngc` to fail for me (fails to resolve dependencies like `rxjs` from Angular code). Normally I have `preserveSymlinks` set to false, but doing that is causing esbuild to fail to resolve child dependencies as mentioned in this issue.

@apesternikov - how did you verify #32  as the fix? I wouldn't mine checking for our code

**Comment by arobinson** [comment]
Think I got the patched file working for the most part, but unfortunately `date-fns` is still causing issues. 

```
✘ [ERROR] [plugin Bazel Sandbox Guard] Failed to resolve import './add/index.js'. Ensure that it is a dependency of an upstream target

    ../../../../../../../../execroot/nw-client/node_modules/.pnpm/date-fns@2.28.0/node_modules/date-fns/esm/index.js:2:31:
      2 │ export { default as add } from "./add/index.js";
        ╵                                ~~~~~~~~~~~~~~~~

✘ [ERROR] [plugin Bazel Sandbox Guard] Failed to resolve import './addBusinessDays/index.js'. Ensure that it is a dependency of an upstream target

    ../../../../../../../../execroot/nw-client/node_modules/.pnpm/date-fns@2.28.0/node_modules/date-fns/esm/index.js:3:43:
      3 │ ...t { default as addBusinessDays } from "./addBusinessDays/index.js";
        ╵                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I'd absolutely love to get rid of date-fns and angular-calendar, but unfortunately we current need them. But #32 did fix the other resolution problems that I was seeing

**Comment by gregmagolan** [comment]
`preserveSymlinks` is now off by default on HEAD which resolves the transitives. However it creates a new issue of esbuild leaving the sandbox. That is detailed here https://github.com/aspect-build/rules_esbuild/issues/58.
